File: src/main/kotlin/com/kotlinbcn/kotlin/java_interop/JavaInterop.java
package com.kotlinbcn.kotlin.java_interop;

import java.util.function.Function;

public class JavaInterop {

    public static void main(final String[] args) {
        final PrinterKotlin printer = new PrinterKotlin();

        printer.print("Jose Luis Franconetti");
        printer.print(() -> "Jose Luis Franconeti");
    }
}

File: src/main/kotlin/com/kotlinbcn/kotlin/java_interop/java_interop.kt
package com.kotlinbcn.kotlin.java_interop

class PrinterKotlin {
    fun print(value: String) = println(value)

    fun print(f: () -> String) = println(f())
}

fun main(args: Array<String>) {
    val printerJava = PrinterJava()

    printerJava.print("Jose Luis Franconetti")
    printerJava.print { "Jose" }
}